cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py

```diff
@@ -1058,9 +1058,9 @@ def get_node_name():
     elif is_state('endpoint.openstack.ready'):
         cloud_provider = 'openstack'
     if cloud_provider == 'aws':
-        return getfqdn()
+        return getfqdn().lower()
```

Comment:
I'm not sure about this one. It looks to me like the node name is only lower-cased when no cloud provider is set.

Here's where the hostname gets lower-cased:

https://github.com/kubernetes/kubernetes/blob/c005b9d0ab6cd963abf66a9a12fb8ad5e48121ad/pkg/util/node/node.go#L45-L56

Here's what kubelet does with it:
https://github.com/kubernetes/kubernetes/blob/c005b9d0ab6cd963abf66a9a12fb8ad5e48121ad/pkg/kubelet/kubelet.go#L372-L384

In short: if there's a cloud provider, kubelet will get the node name from that instead of from `GetHostname`.

For the AWS cloud provider, I don't see it being lower-cased:

https://github.com/kubernetes/kubernetes/blob/c005b9d0ab6cd963abf66a9a12fb8ad5e48121ad/pkg/cloudprovider/providers/aws/aws.go#L4294-L4296

I don't actually know if it's possible for AWS's `PrivateDnsName` to contain uppercase characters, but I would like to err on the side of caution and assume it might.

@dshcherb can we reduce the scope of this change to only apply when no `cloud-provider` is set? I would feel safest with something like this instead:
```
if cloud_provider == '':
    return gethostname().lower()
elif cloud_provider == 'aws':
    return getfqdn()
else:
    return gethostname()
```

Reply:
The K8s docs seem to be pretty clear on conventions:

https://kubernetes.io/docs/concepts/overview/working-with-objects/names/
"By convention, the names of Kubernetes resources should be up to maximum length of 253 characters and consist of *lower case alphanumeric characters*, -, and ., but certain resources have *more specific restrictions*."
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/identifiers.md - this has a bit vague references to resources specifically but mentions DNS labels and lowercase nevertheless.

I have also followed some code paths (from "node create" to the request validation) - looks like an uppercase node (resource) name would not pass ValidateNodeName check as it requires a node name to be have the following regex:
const dns1035LabelFmt string = "[a-z]([-a-z0-9]*[a-z0-9])?":
https://gist.github.com/dshcherb/1ba9b9cb62f8753de210bc459115c07a

I could reduce the scope but I don't think this is necessary due to the above.

Thoughts?

Reply:
Thanks @dshcherb. You're right - it looks like the apiserver does adhere strictly to that convention:

```
$ cat test-node.yaml 
apiVersion: v1
kind: Node
metadata:
  name: TEST-NODE

$ kubectl create -f test-node.yaml 
The Node "TEST-NODE" is invalid: metadata.name: Invalid value: "TEST-NODE": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
```

I'm +1 to this as-is then - the charm should always convert it to lower case. Thanks for setting me straight.